File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        
        left = 0 # buying
        right = 1 # selling

        maxPrice = 0

        while right < len(prices):

            # profitable?
            if prices[left] < prices[right]:
                profit = prices[right] - prices[left]
                maxPrice = max(maxPrice, profit)
            else: 
                left = right # if right < left then we have a new min buy
            right += 1

        return maxPrice

File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int left = 0;
        int right = heights.size()-1;
        int best = 0;

        while (right > left) {
            int area = min(heights[right], heights[left]) * (right-left);
            best = max(area, best);
            if (heights[left] <= heights[right]) 
                left++;
            else
                right--;
        }
        return best;
    }
};
